packages/app/components/EditTerritory/UploadButton/decorators/withUploadShapefile.ts

```diff
@@ -0,0 +1,70 @@
+import { compose, withHandlers } from 'recompose';
+import * as shapefile from 'shapefile';
+import { subscribe } from 'react-contextual';
+
+import {
+  UPLOADED_TERRITORY,
+  WithEditTerritoryStoreProps
+} from '../../decorators/withEditTerritoryStore';
+
+export interface WithUploadShapefileProps {
+  handleUploadShapefile: (event: React.ChangeEvent<HTMLInputElement>) => void;
+}
+
+/**
+ * Read an uploaded file via <input type="file" /> as an ArrayBuffer.
+ *
+ * @param inputFile - The uploaded file.
+ * @see https://blog.shovonhasan.com/using-promises-with-filereader/
+ */
+const readAsArrayBuffer = (inputFile: File): Promise<ArrayBuffer> => {
+  const reader = new FileReader();
+  return new Promise((resolve, reject) => {
+    reader.onerror = () => {
+      reader.abort();
+      reject(new Error('Problem parsing shapefile.'));
+    };
+
+    reader.onload = () => {
+      resolve(reader.result as ArrayBuffer);
+    };
+    reader.readAsArrayBuffer(inputFile);
+  });
+};
+
+/**
+ * Decorator to add a handler when we update the polygon we draw on the map.
+ */
+export default compose(
+  subscribe('withEditTerritoryStore'),
+  withHandlers<WithEditTerritoryStoreProps, {}>({
+    handleUploadShapefile: ({ addShapefile }) => async ({
+      target: { files }
+    }: React.ChangeEvent<HTMLInputElement>) => {
+      try {
+        const uploaded = files[0];
```

Comment:
Any logic to check mime type or file extension?

Reply:
Unfortunately the mime type is empty when I upload the shapefiles given by the mappers.